course5/homework06/ustring.h

```diff
@@ -0,0 +1,121 @@
+#pragma once
+
+#include <iostream>
+#include <stdexcept>
+#include <limits>
+#include <memory>
+#include <cassert>
+
+namespace Course {
+
+class String
+{
+public:
+    static constexpr std::size_t npos = -1;
+
+    String() = default;
+
+    String(const char *str) { *this = str; }
+    String(const char *str, size_t length);
+    String(String const &str) { *this = str; }
+    String(String &&rhs);
+
+    String& operator=(const char *str);
+    String& operator=(String const &rhs);
+    String& operator=(String &&rhs);
+
+    // Element access
+    char& operator[](std::size_t pos) { return data()[pos]; }
+
+    const char& operator[](std::size_t pos) const { return data()[pos]; }
+
+    char& at(std::size_t pos)
+    {
+        if (pos >= mLength)
+            throw std::out_of_range("String::at");
+        return (*this)[pos];
+    }
+
+    const char& at(std::size_t pos) const
+    {
+        if (pos >= mLength)
+            throw std::out_of_range("String::at");
+        return (*this)[pos];
+    }
+
+    char& front() { return (*this)[0]; }
+    const char& front() const { return (*this)[0];  }
+
+    char& back() { return (*this)[mLength-1]; }
+    const char& back() const { return (*this)[mLength-1]; }
+
+    char* data()
+    {
+        if (length() == 0)
+            return emptyString;
```

Comment:
Well, that explains why I had to remove a check when I changed the code to work on C++14. I was wondering why the data() returns the same pointer after coping an std::string, but too sleepy to look at it closely. I pre gcc 5 std::string is copy on write, and data() is const so copy constructor doesn't actually have to copy anything at all.